File: backend/routes/slots.js
import express from 'express';
import * as slotService from '../services/slotService.js';
import { validateSlotUpdate } from '../middleware/validator.js';
import { requireOperator, requireAdmin } from '../middleware/auth.js';


const router = express.Router();

router.get('/', requireOperator, async (req, res, next) => {
  try {
    const slots = await slotService.getAllSlots();
    res.json(slots);
  } catch (err) {
    next(err);
  }
});

router.get('/:id', requireOperator, async (req, res, next) => {
  try {
    const slot = await slotService.getSlotById(req.params.id);
    if (!slot) return res.status(404).json({ error: 'Slot not found' });
    res.json(slot);
  } catch (err) {
    next(err);
  }
});

router.post('/update', requireAdmin, validateSlotUpdate, async (req, res, next) => {
  try {
    const { controllerId, status } = req.body;
    const slot = await slotService.updateSlotStatus(controllerId, status);
    
    // We should also emit socket update here if triggered manually
    if (req.app.get('io')) {
      req.app.get('io').emit('slotUpdated', slot);
    }
    
    res.json(slot);
  } catch (err) {
    next(err);
  }
});

export default router;

File: backend_py/app/routes/slots.py
from fastapi import APIRouter, Depends, HTTPException, Request
from pydantic import BaseModel
from app.services import slot_service
from app.middleware.auth import require_operator, require_admin
from typing import List, Dict, Any

router = APIRouter(prefix="/api/slots", tags=["slots"])

class UpdateSlotModel(BaseModel):
    controllerId: str
    status: str

@router.get("/")
async def get_slots(operator=Depends(require_operator)):
    try:
        slots = await slot_service.get_all_slots()
        return slots
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/{id}")
async def get_slot(id: int, operator=Depends(require_operator)):
    try:
        slot = await slot_service.get_slot_by_id(id)
        if not slot:
            raise HTTPException(status_code=404, detail="Slot not found")
        return slot
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/update")
async def update_slot(data: UpdateSlotModel, request: Request, admin=Depends(require_admin)):
    try:
        slot = await slot_service.update_slot_status(data.controllerId, data.status)
        
        # Socket emission
        sio = getattr(request.app.state, "sio", None)
        if sio:
            await sio.emit("slotUpdated", slot)
            
        return slot
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
